### ISSUE TITLE
pmpro_visit Cookie is killing caching

### ISSUE THREAD

**Issue by megaz555** [strangerstudios/paid-memberships-pro#1158]
**Describe the bug**
The pmpro_visit cookie is enabled for every visit, which breaks caching plugins. The code in login.php file under pmpro_report_track_values previously used to do this:

if(!is_user_logged_in()
return false;

This made caching plugins work perfectly fine, since the cookie was not being set. This code has been removed and now the plugin is setting the pmpro_visit cookie for every load, which breaks all caching plugins. If I manually add the above code into the file, caching is back to normal. Please re-instate this code, so that the cookie is not used for non-logged in users.

Thank you.

**To Reproduce**
Use curl to crawl a page and you will see the pmpro_visit cookie.

**Expected behavior**
Cookies should not be initialized for non-logged in users

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only Paid Memberships Pro plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Memberlite](https://wordpress.org/themes/memberlite/).
- [x] I can reproduce this bug consistently using the steps above.

**Comment by ideadude** [comment]
Thank you for bringing this up and looking into this.

I am not sure how the cookie being set was breaking caching. I assume the cache would work and the cookie just wouldn't get updated.

I would consider changing this behavior. Tracking visits, especially for non members, is not a core feature of the plugin. Still some people might rely on that a bit.

In the meantime, here is some code that will deactivate the visit tracking for non users:
https://gist.github.com/ideadude/27912cd472e1216ba73420386ddae2ce

Let me know if that works. If you use this custom code, instead of patching PMPro, you will still be able to upgrade the plugin without removing your edits.

**Comment by ideadude** [comment]
> I am not sure how the cookie being set was breaking caching. I assume the cache would work and the cookie just wouldn't get updated.

BTW, here I am not saying that I believe the cookies are not breaking cache. I am admitting that I don't know what kind of cache or how that is happening, and I am interested in more information about this.

In general, moving cookie activity to frontend JavaScript is beneficial for a number of a reasons and would probably fix this issue. That's a bit more work though, but something we can look into.

**Comment by charlesfsl** [comment]
We have a site hosted with https://pantheon.io/ where this cookie's name is breaking their Varnish cache by making requests from anyone with this cookie un-cacheable. Their support pointed us to [this tool](https://varnishcheck.pantheon.io/varnish) for checking Varnish on the site and told us the cookie's name is the reason for the issue. They indicated that changing the cookie's name was all that was needed from their end for resolving this issue, but I do not know the implications of such a change within PM Pro. Thanks!

**Comment by ideadude** [comment]
I've confirmed that PMPro is creating a PHPSESSID cookie even when we wouldn't need it. I'm trying to figure out why that is the case.

I've also confirmed that activating the WP Native Sessions plugin by Pantheon does change the name of the PHPSESSID cookie and might avoid issues with Varnish cache. https://wordpress.org/plugins/wp-native-php-sessions/

**Comment by ideadude** [comment]
We should only try to use sessions on the checkout page, but I noticed we were setting sessions on every page load.

I found that our recaptcha code is checking for the recaptcha validation. For now, here is a commit that stops that if you aren't using the recaptcha setting. Ideally we would want to make sure we are on the checkout page too. That's not an easy check because people can insert checkout shortcodes and blocks into any page, but before release we will want to add some logic to make sure that the check only happens on pages that could have recptchas. We can also look into using something other than a session id to store the validation, but I'm pretty sure we introduced this to support Stripe payment intents and can't be avoided without creating our own session like thing which I won't do.

To recap, there are 2 fixes which should work now.

1. Use the WP Native PHP Sessions plugin.
2. Apply this new commit and disable recaptcha.

We will have an update soon that will only load the session when visitors hit the checkout page (or something otherwise triggers a session).

**Comment by ideadude** [comment]
Here's the patch to only load recaptcha on the checkout page.

https://github.com/strangerstudios/paid-memberships-pro/commit/3dc95acb06e131efca128208b6a22ec0f77f6a1d

On a clean install, I have no session until I hit the checkout page. Then I have a session. After that point, Varnish cache set up to ignore users with a session cookie will not work for me anymore that session. That's the best we can do. You can use the WP Native PHP Sessions plugin, which might avoid this cache issue.

**Comment by ideadude** [comment]
If the pmpro_visit cookie is breaking your cache, set up your cache to not break when it sees that cookie or use our gist to disable that tracking.

**Comment by charlesfsl** [comment]
Hi @ideadude thanks for your work on this. I can try out the patch when the work week starts back. 

For what it's worth, we've been running the WP Native PHP Sessions plugin for months now and it has not resolved the issue. We also tested the site with and without the WP Native PHP Sessions plugin and that had no effect on the pmpro cookie issue.

We're using the default config for Pantheon's Varnish cache, which we get little ability to modify. [The extent to which we can is described in this doc I linked to in my previous comment.](https://pantheon.io/docs/caching-advanced-topics) We cannot set it up to avoid this cookie breaking the cache; it's impossible without buying Pantheon.

I think reducing this cookie's session to the checkout page is a reasonable compromise. We'll let you know how it goes!

**Comment by ideadude** [comment]
Here are the cookies Pantheon is looking for:

````
NO_CACHE
S+ESS[a-z0-9]+
fbs[a-z0-9_]+
SimpleSAML[A-Za-z]+
PHPSESSID
wordpress[A-Za-z0-9_]*
wp-[A-Za-z0-9_]+
comment_author_[a-z0-9_]+
duo_wordpress_auth_cookie
duo_secure_wordpress_auth_cookie
bp_completed_create_steps # BuddyPress cookie used when creating groups
bp_new_group_id # BuddyPress cookie used when creating groups
wp-resetpass-[A-Za-z0-9_]+
(wp_)?woocommerce[A-Za-z0-9_-]+
amazon_Login_[A-Za-z0-9_]+
````

With that plugin activated, it creates cookies like SSESS010406e995291c043703177856ee2e84, which match the second rule there. So that plugin won't avoid the cache issues as I thought. My bad.

I don't believe the pmpro_visit cookie fits any of those rules, and should be fine.

**Comment by megaz555** [comment]
What's the reason for wanting to insert a cookie on every page load of _every_ page? No other plugin does that. Can't you just skip all sessions and cookies until a person starts the checkout process?

By the way, the cookie is inserted via the file adminpages -> reports -> login.php, not the recaptcha file that you patched.

Someone from your team responded saying that this functionality will not be removed:
https://wordpress.org/support/topic/pmpro_visit-cookie-for-non-logged-in-users/

I am baffled that you guys are doing this. Your plugin gets installed, and the entire site cannot get cached because you are pushing a cookie / starting a session. Do you realize that you are making the Internet slower? There are so many site owners that would have no clue that their servers are getting hammered with origin requests, even while sitting behind caching servers.

Very disappointing. I hope you can fix this, once and for all.

**Comment by joshualibrarian** [comment]
Neither that patch for the captcha file, nor even commenting out the only `set_cookie()` in the entire codebase (in `adminpages/reports/login.php`) seems to turn that cookie off (regardless of the WP Native Sessions plugin).  Where the is that SSES[etc] cookie being set?  When I deactivate PMPro, it's gone, so it must be in here somewhere.  :/

**Comment by andrewlimaza** [comment]
@megaz555 please update to version 2.4.1 and disable the pmpro_visit cookie using this code snippet: https://gist.github.com/andrewlimaza/12dc2f179aea63960e54cc0df30e18e0

I did a test using the Proxy Cache Purge plugin to test the cookies/caching, here's a link to the plugin - https://wordpress.org/plugins/varnish-http-purge/

This was run on a 'bare-minimum' setup which included the latest version of WordPress, Paid Memberships Pro 2.4.1 and this Proxy Cache Purge plugin. Let me know if this helps at all - if it's still creating the PHPSESSID cookie on all pages, please reach out to your host to ensure that the server-side caching is cleared entirely.

**Comment by joshualibrarian** [comment]
@andrewlimaza Thanks for that.  It seems that just updating to 2.4.1 fixed the issue, and the code snippet wasn't required.  Varnish cache is now functional!  :)

**Comment by andrewlimaza** [comment]
Thanks for the feedback @joshualibrarian, so glad to hear it's working again. :)

**Comment by insomniacsociety** [comment]
Any further updates on this? I'm using 2.5.9.1 and I'm still having this issue where PMPro is preventing varnish from caching on Pantheon. I'm using their Native Sessions plugin and I've also tried using Andrew's remove-pmpro-visits-reporting.php snippet. I'd rather not try modifying code in the plugin but I'm close to my wit's end here.

**Comment by ideadude** [comment]
I'm going to close this issue because I think we've fixed this specific issue integrating with Varnish/Pantheon.

@insomniacsociety if you are still having trouble, please reach out through our paid support or post a new issue with more details on the issue you are having.
